### ISSUE TITLE
Resolving an error with code kCFURLErrorCancelled does not resolve

### ISSUE THREAD

**Issue by Slessi** [mxcl/PromiseKit#613]
* PromiseKit version: 4.0.4
* Xcode version: 8.2.1 (8C1002)
* CocoaPods version: 1.1.1

When using `promiseDataTaskWithRequest` and I cancel the relevant `NSURLSessionTask` before it completes, my promises do not resolve.

After some digging I was able to easily replicate the issue I experience as follows:

resolve a [kCFURLErrorBadServerResponse](https://developer.apple.com/reference/cfnetwork/cfnetworkerrors/kCFURLErrorBadServerResponse) error code, `caught` logged.
```objective-c
[AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
    resolve([NSError errorWithDomain:NSURLErrorDomain code:kCFURLErrorBadServerResponse userInfo:nil]);
}].catch(^{
    NSLog(@"caught");
}).then(^{
    NSLog(@"not caught");
});
```

resolve a [kCFURLErrorCancelled](https://developer.apple.com/reference/cfnetwork/cfnetworkerrors/kCFURLErrorBadServerResponse) error code, nothing is logged (promise does not resolve at all)
> Note the `code` change to `kCFURLErrorCancelled`
```objective-c
[AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
    resolve([NSError errorWithDomain:NSURLErrorDomain code:kCFURLErrorCancelled userInfo:nil]);
}].catch(^{
    NSLog(@"caught");
}).then(^{
    NSLog(@"not caught");
});
```

**Comment by mxcl** [comment]
This is expected behavior, please read the documentation on cancellation: http://promisekit.org/docs/

**Comment by Slessi** [comment]
Seems unintuitive to me but thanks for pointing me in the right direction

**Comment by mxcl** [comment]
Most of the time it is a very useful feature, hence its existence despite it being confusing for people who jumped straight into PromiseKit without reading our very minimal documentation first.

**Comment by mxcl** [comment]
Just for clarification, you **cancelled** the url request, what were you expecting to happen?

**Comment by Slessi** [comment]
I read your documentation  when I started using your library, but this was many many months ago now, you can't actually expect me to remember little nuances like this.

I am coming from JavaScript where this would be a violation of the A+ promises spec I am used to in which a promise's state is either pending, fulfilled or rejected, whereas this is some weird 4th state between pending and fulfilled.

But anyway this is not JavaScript and its your library so do as you please. My use case was not as simple as cancelling a request, I just cut straight to the point for the purposes of simplifying the issue.

I am implementing certificate pinning on my app. If you fail the auth challenge, iOS by default cancels the task. There is no special error code for it, just gets marked as cancelled. Pretty silly of Apple to not use an appropriate error code for this case but it is what it is and I couldn't find a way of ending the task with a custom error.

tl;Dr I was expecting my catch function to run whrn my task was cancelled. I am on my phone or I'd link you to the relevant methods I am referring to for certificate pinning

**Comment by mxcl** [comment]
> I read your documentation when I started using your library, but this was many many months ago now, you can't actually expect me to remember little nuances like this.

Yeah, I get you. Sorry for my snide response.

> I am coming from JavaScript where this would be a violation of the A+ promises spec I am used to in which a promise's state is either pending, fulfilled or rejected, whereas this is some weird 4th state between pending and fulfilled.

The promise *is* rejected. But we have a sub-state for rejection.

> But anyway this is not JavaScript and its your library so do as you please.

Conforming to A+ is important to this library.

> I am implementing certificate pinning on my app. If you fail the auth challenge, iOS by default cancels the task. There is no special error code for it, just gets marked as cancelled. Pretty silly of Apple to not use an appropriate error code for this case but it is what it is and I couldn't find a way of ending the task with a custom error.

Yeah, this makes sense. So indeed, it was not *you* asking for cancellation it was some decision on Apple's behalf that made behavior from PMK unexpected.

I'm going to reopen this, as a) cancellation should be explicit, so we should figure out how to at least post a debug warning for this case and b) many of the reasons I had for initially adding this behavior are less compelling to me a few years of PMK use down the line.

Really appreciate your report.

**Comment by mxcl** [comment]
In PMK5 a message is logged whenever a cancellation message is ignored by a `catch` handler. I played with removing cancellation altogether, but in all my apps it is too useful to be able to `reject` a chain without it leading to a `catch` handling the error. Thoughts welcome!

**Comment by Slessi** [comment]
@mxcl I do agree that the option to end a promise chain without it leading to a catch can be useful, maybe as a seperate function like `abort`, and having an option to automatically `abort` a promise chain in the event of cancellation can be useful also to some people.

What I do not agree with is that by *default* any `reject` (for whatever reason) may not bubble down to a `catch` because of another user's preferences. Even if that may be the option everyone secretly wants.

I think that as it goes against the expected default behaviour of `reject` it should be opt-in, not opt-out.

**Comment by mxcl** [comment]
Your abort idea would still conflict with your primary concern though wouldn't it? That any user may cause a chain to not have a catch handled. Or did I miss the point?

**Comment by Slessi** [comment]
It wouldn't because would be a distinction between an `abort` and a `reject`.

`resolve([NSError errorWithDomain:NSURLErrorDomain code:kCFURLErrorCancelled userInfo:nil]);`
Would bubble to the next catch

`resolve([AnyPromise abort])` (or something similar)
Would act as it currently does when resolving a cancellation

Now, a `reject` will *always* bubble up to the next `catch`, which IMO is the expected behaviour unless you specifically state otherwise by making an `abort` or turning on the abort all cancellations feature with a flag like `[PromiseKit setAutomaticallyAbortsCancellation:YES]`.

My primary concern was just the confusing nature of having rejected promises that dont trigger a `catch` and having the cancellation feature enabled by default without the user being able to know unless he remembers the documentation. The logging does resolve the user not knowing part though.

**Comment by Slessi** [comment]
You're focusing on the wrong bit, the abort/reject thing is just semantics to distinguish between something which leads to a catch and something which does not.

From your docs
> Cancelled promises are *still* rejected promises

Even if `p. isRejected ` returns `true`, it hasn't behaved like a rejected promise unless it triggered a catch.

---

The confusion is that an `abort` happens without informing the user. This is caused by it being both non-standard behaviour and also the default option. If it must remain this way, I would agree `NSLog` is the next best step.

Otherwise, my proposed solution is to only `abort` cancellations if the user manually enables it.

Examples of what I am talking about to make it non-default

* `[PromiseKit setAutomaticallyAbortsCancellation:YES]`
* `[PromiseKit setDefaultCatchPolicy:PMKCatchPolicyAllErrors]`
* Change the default implementation of `catch` to use `PMKCatchPolicyAllErrors` instead of `PMKCatchPolicyAllErrorsExceptCancellation`

I'm sure there are many other choices also.

**Comment by mxcl** [comment]
Canceled errors log since PMK 5.
